crates/codegen/parser/runtime/src/support/recovery.rs

```diff
@@ -33,4 +34,194 @@ impl ParserResult {
             result => result,
         }
     }
+
+    pub fn recover_until_with_nested_delims(
+        self,
+        input: &mut ParserContext,
+        next_token: impl Fn(&mut ParserContext) -> Option<TokenKind>,
+        leading_trivia: impl Fn(&mut ParserContext) -> ParserResult,
+        expected: TokenKind,
+        delims: &[(TokenKind, TokenKind)],
+    ) -> ParserResult {
+        let before_recovery = input.position();
+
+        // Parse trivia so that we can compare the next relevant token to the expected
+        let leading_trivia = {
+            let start = input.position();
+            if let ParserResult::Match(leading_trivia) = leading_trivia(input) {
+                leading_trivia.nodes
+            } else {
+                input.set_position(start);
+                vec![]
+            }
+        };
+
+        let start = input.position();
+
+        let mut peek_token = || {
+            let start = input.position();
+            let tok = next_token(input);
+            input.set_position(start);
+            tok
+        };
+
+        match self {
+            ParserResult::IncompleteMatch(mut result) => {
+                result.nodes.extend(leading_trivia);
```

Comment:
The 4 recovery branches are the same - I haven't collapsed it yet while working because I wasn't sure if the logic is going to stay the same. Happy to collapse/clean this up if the approach here is considered good enough

Reply:
Cleaned up in 75fa300301a3a267efce0e146fa8122305a85c9c